src/lib.rs

```diff
@@ -272,23 +271,26 @@ impl<T: Ord> OrderedCollection<T> {
     /// s.insert(89);
     /// s.insert(7);
     /// s.insert(12);
-    /// let a = OrderedCollection::from_sorted_iter(s.iter());
-    /// assert_eq!(a.find_gte(50), Some(&&89));
+    /// let a = OrderedCollection::from_sorted_iter(s.iter().copied());
+    /// assert_eq!(a.find_gte(50), Some(&89));
     /// ```
     ///
     pub fn from_sorted_iter<I>(iter: I) -> Self
     where
         I: IntoIterator<Item = T>,
-        I::IntoIter: ExactSizeIterator<Item = T>,
+        I::IntoIter: ExactSizeIterator,
     {
         let iter = iter.into_iter();
         let n = iter.len();
-        let mut context = (Vec::with_capacity(n), iter);
-        eytzinger_walk(&mut context, 0);
+        // the root element is located in a right slot of the root node, so we need to allocate n + 1 elements
+        let mut context = (Vec::with_capacity(n + 1), iter);
+        // Here we start filling the vector from the lement with index 1, leaving 0th element uninitialized.
+        // Therefore accessing the 0th element is Undefined Behavior.
+        eytzinger_walk(&mut context, 1);
         let (mut items, _) = context;
 
-        // it's now safe to set the length, since all `n` elements have been inserted.
-        unsafe { items.set_len(n) };
+        // it's now safe to set the length, since all `n` elements have been inserted (do not forget the root node).
+        unsafe { items.set_len(n + 1) };
```

Comment:
I don't think this is right, since `[0]` has _not_ been inserted and thus holds an invalid value. If we nevertheless include it in the resulting vector, then `[0]` would become accessible from safe code, which isn't sound.

Reply:
I see 2 possible solution:

1. put a `T::default()` in `[0]` which requires `T: Default`
2. do something like `items[0] = items[1]` after eytzinger constructed, which requires `T: Copy` and `iter.len() > 0`.

It seems to me that 1 is more general solution. What do you think?

P.S. doesn't the fact that we never read `[0]` make this code sound? Or you talking more about soundness in the context of future changes?

Reply:
Quoth the safety docs for [`Vec::set_len`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len):

> The elements at `old_len..new_len` must be initialized.

Part of this is because when the `Vec` is dropped, it will in turn drop each element, which includes `[0]`, but trying to drop an uninitialized value is UB.

One option here would be to store a `Vec<MaybeUninit<T>>` instead. When we slice into the `Vec` and get out `&[MaybeUnint<T>]` we can then cast that to `&[T]` (as long as the index is `> 0`), and it's entirely a compile-time operation so no runtime penalty. To make that work we'd also need to `impl Drop` and have that do:
```rust
let items: &mut Vec<T> = unsafe { mem::transmute(&mut self.items) };
items.truncate(1);
```
In other words, it'll `truncate` all elements except `[0]` under the type `T`, which means it'll call `T::drop` for those. The `[0]` is left in there for when the `Vec` is dropped, which is fine, because it's then `MaybeUninit<T>`, so `T::drop` won't be called for that one.

Reply:
Indeed, it makes sense. Although, it is interesting that miri doesn't complain about it at all. Maybe it is because we're only testing `OrderedCollection` for primitive types that are `Copy` and do not require dropping.

Reply:
Also, here is another perspective...

It might be a good call to require `T: Copy + 'static` by default. For general kind of data structure, this is quite limiting. But if you see `OrderedCollection` as a specialized performance-oriented data structure I don't think there are good reasons to use it with references. You are throwing away significant portion of the performance. This is what basically happens in #7

Reply:
It's an interesting question — I genuinely don't know whether this will be faster than the alternative for some `Copy` or non-`'static` types. I'm inclined to not _require_ those bounds unless we have strong evidence that that's the case though. Although we can spell it out in the docs!

Re miri: yeah, that's almost certainly what happens — worth trying running miri over a non-`Copy` `T` type and see if it complains then.

Reply:
Miri test added, `MaybeUninit` solution implemented

Reply:
I would probably avoid doing the transmute of the whole `Vec` in the walk fn, and instead only `MaybeUninit::assume_init` when we access individual elements. Transmuting the whole thing increases the risk that someone accidentally tries to access `[0]` without realizing that's UB.

Also, safety comments please ⚠️

Reply:
Please, check again.